File: JT3DScrollView/JT3DScrollView.m
//
//  JT3DScrollView.m
//  JT3DScrollView
//
//  Created by Jonathan Tribouharet
//

#import "JT3DScrollView.h"

#define RADIANS_TO_DEGREES(radians) ((radians) * (180.0 / M_PI))
#define DEGREES_TO_RADIANS(angle) ((angle) / 180.0 * M_PI)

@implementation JT3DScrollView

- (instancetype)initWithFrame:(CGRect)frame
{
    self = [super initWithFrame:frame];
    if(!self){
        return nil;
    }
    
    [self commonInit];
    
    return self;
}

- (id)initWithCoder:(NSCoder *)aDecoder
{
    self = [super initWithCoder:aDecoder];
    if(!self){
        return nil;
    }
    
    [self commonInit];
    
    return self;
}

- (void)commonInit
{
    self.pagingEnabled = YES;
    self.clipsToBounds = NO;
    self.showsHorizontalScrollIndicator = NO;
    self.showsVerticalScrollIndicator = NO;
    
    self.effect = JT3DScrollViewEffectNone;
}

- (void)setEffect:(JT3DScrollViewEffect)effect
{
    self->_effect = effect;
    
    switch (effect) {
        case JT3DScrollViewEffectTranslation:
            self.angleRatio = 0.;
            
            self.rotationX = 0.;
            self.rotationY = 0.;
            self.rotationZ = 0.;
            
            self.translateX = .25;
            self.translateY = .25;
            break;
        case JT3DScrollViewEffectDepth:
            self.angleRatio = .5;
            
            self.rotationX = -1.;
            self.rotationY = 0.;
            self.rotationZ = 0.;
            
            self.translateX = .25;
            self.translateY = 0.;
            break;
        case JT3DScrollViewEffectCarousel:
            self.angleRatio = .5;
            
            self.rotationX = -1.;
            self.rotationY = 0.;
            self.rotationZ = 0.;
            
            self.translateX = .25;
            self.translateY = .25;
            break;
        case JT3DScrollViewEffectCards:
            self.angleRatio = .5;
            
            self.rotationX = -1.;
            self.rotationY = -1.;
            self.rotationZ = 0.;
            
            self.translateX = .25;
            self.translateY = .25;
            break;
        default:
            self.angleRatio = 0.;
            
            self.rotationX = 0.;
            self.rotationY = 0.;
            self.rotationZ = 0.;
            
            self.translateX = 0.;
            self.translateY = 0.;
            break;
    }
    
    [self setNeedsDisplay];
}

- (void)layoutSubviews
{
    [super layoutSubviews];
    
    CGFloat contentOffsetX = self.contentOffset.x;
    CGFloat contentOffsetY = self.contentOffset.y;
    
    for(UIView *view in self.subviews){
        CATransform3D t1 = view.layer.transform; // Hack for avoid visual bug
        view.layer.transform = CATransform3DIdentity;
        
        CGFloat distanceFromCenterX = view.frame.origin.x - contentOffsetX;
//        CGFloat distanceFromCenterY = view.frame.origin.y - contentOffsetY;
        
        view.layer.transform = t1;
                
        distanceFromCenterX = distanceFromCenterX * 100. / CGRectGetWidth(self.frame);
//        distanceFromCenterY = distanceFromCenterY * 100. / CGRectGetHeight(self.frame);
        
        CGFloat angle = distanceFromCenterX * self.angleRatio;
        
        CGFloat offset = distanceFromCenterX;
        CGFloat translateX = (CGRectGetWidth(self.frame) * self.translateX) * offset / 100.;
        CGFloat translateY = (CGRectGetWidth(self.frame) * self.translateY) * abs(offset) / 100.;
        CATransform3D t = CATransform3DMakeTranslation(translateX, translateY, 0.);

        view.layer.transform = CATransform3DRotate(t, DEGREES_TO_RADIANS(angle), self.rotationX, self.rotationY, self.rotationZ);
    }
}

- (NSUInteger)currentPage
{
    CGFloat pageWidth = self.frame.size.width;
    float fractionalPage = self.contentOffset.x / pageWidth;
    return lround(fractionalPage);
}

- (void)loadNextPage:(BOOL)animated
{
    [self loadPageIndex:self.currentPage + 1 animated:animated];
}

- (void)loadPreviousPage:(BOOL)animated
{
    [self loadPageIndex:self.currentPage - 1 animated:animated];
}

- (void)loadPageIndex:(NSUInteger)index animated:(BOOL)animated
{
    CGRect frame = self.frame;
    frame.origin.x = frame.size.width * index;
    frame.origin.y = 0;
    
    [self scrollRectToVisible:frame animated:animated];
}

@end

File: Source/JT3DScrollView.swift
//
//  JT3DScrollView.swift
//  JT3DScrollView
//
//  Created by Jonathan Tribouharet
//

import UIKit

public enum JT3DScrollViewEffect {
	case none, translation, depth, carousel, cards
}

open class JT3DScrollView: UIScrollView {

	open var angleRatio: CGFloat = 0
	open var rotationX: CGFloat = 0
	open var rotationY: CGFloat = 0
	open var rotationZ: CGFloat = 0
	open var translateX: CGFloat = 0
	open var translateY: CGFloat = 0

	open var effect: JT3DScrollViewEffect = .none {
		didSet {
			loadEffectValues()
		}
	}

	open var currentPage: Int {
		get {
			let pageWidth = self.frame.size.width
			if pageWidth <= 0 {
				return 0
			}

			let fractionalPage = max(self.contentOffset.x, 0) / pageWidth
			return Int(round(fractionalPage))
		}
	}

	public override init(frame: CGRect) {
		super.init(frame: frame)
		commonInit()
	}

	public required init?(coder aDecoder: NSCoder) {
		super.init(coder: aDecoder)
		commonInit()
	}

	open func commonInit() {
		isPagingEnabled = true
		clipsToBounds = false
		showsHorizontalScrollIndicator = false
		showsVerticalScrollIndicator = false
	}

	open func loadEffectValues() {
		switch effect {
			case .translation:
			angleRatio = 0
			rotationX = 0
			rotationY = 0
			rotationZ = 0
			translateX = 0.25
			translateY = 0.25
			case .depth:
			angleRatio = 0.5
			rotationX = -1
			rotationY = 0
			rotationZ = 0
			translateX = 0.25
			translateY = 0
			case .carousel:
			angleRatio = 0.5
			rotationX = -1
			rotationY = 0
			rotationZ = 0
			translateX = 0.25
			translateY = 0.25
			case .cards:
			angleRatio = 0.5
			rotationX = -1
			rotationY = -1
			rotationZ = 0
			translateX = 0.25
			translateY = 0.25
			case .none:
			angleRatio = 0
			rotationX = 0
			rotationY = 0
			rotationZ = 0
			translateX = 0
			translateY = 0
		}

		setNeedsDisplay()
	}

	open func loadNextPage (animated: Bool) {
		loadPageIndex(index: currentPage + 1, animated: animated)
	}

	open func loadPreviousPage (animated: Bool) {
		loadPageIndex(index: currentPage - 1, animated: animated)
	}

	open func loadPageIndex(index: Int, animated: Bool) {
		var newFrame = self.frame
		newFrame.origin.x = newFrame.size.width * CGFloat(index)
		newFrame.origin.y = 0

		scrollRectToVisible(newFrame, animated: animated)
	}

	open override func layoutSubviews () {
		super.layoutSubviews()

		let contentOffsetX = self.contentOffset.x
		for view in subviews {

			let t1 = view.layer.transform // Hack for avoid visual bug
			view.layer.transform = CATransform3DIdentity

			var distanceFromCenterX = view.frame.origin.x - contentOffsetX

			view.layer.transform = t1

			distanceFromCenterX = distanceFromCenterX * 100 / self.frame.width

			let angle = distanceFromCenterX * self.angleRatio

			let offset = distanceFromCenterX;
			let translateX = (self.frame.width * self.translateX) * offset / 100
			let translateY = (self.frame.width * self.translateY) * fabs(offset) / 100
			let t = CATransform3DMakeTranslation(translateX, translateY, 0)

			view.layer.transform = CATransform3DRotate(t, (angle * .pi / 180), self.rotationX, self.rotationY, self.rotationZ)
		}
	}

}
